File: in_Python/0238 Product of Array Except Self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        left = []
        left.append(1)
        for i in range(1,len(nums)):
            left.append(left[i-1] * nums[i-1])
            
        right = []
        for i in range(len(nums)-1): right.append(0)
        right.append(1)

        for i in range(len(nums)-2,-1,-1):
            right[i] = right[i+1] * nums[i+1]
        res = []
        for i in range(len(left)):
            res.append(left[i] * right[i])
        return res

File: in_Java/0238 Product of Array Except Self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] res = new int[nums.length];
        res[res.length-1] = 1;
        for(int i = res.length-2; i >= 0; i--){
            res[i] = res[i+1] * nums[i+1];
        }
        int t = nums[0];
        nums[0] = 1;
        for(int i = 1; i < nums.length; i++){
            int tt = nums[i];
            nums[i] = t;
            t = t * tt;
        }
        for(int i = 0; i < res.length; i++){
            res[i] = res[i] * nums[i];
        }
        return res;
    }
}